### PR TITLE
CI: added updated winget package submission

### LINKED ISSUE #7808: Add Deskflow to winget packages repository
### Feature Request

Since I started using UniGetUI for managing most of my windows updates, it would be veri cool to get also deskflow updates through winget package manager. 

If needed I can create the first starting manifest with the latest release, but to keep the package updated, it requirese or a manual submission at every update or an automated CI that does that([winget-updater](https://github.com/marketplace/actions/winget-updater) should do the job).

### PR DESCRIPTION
**Pull request by nicola02nb**
resolve #7808 
Took the idea from: https://techcommunity.microsoft.com/t5/educator-developer-blog/wingetcreate-keeping-winget-packages-up-to-date/ba-p/4037598

### CODE REVIEW

**Comment by nicola02nb** [comment]
Should check if this secret works with the PR submission generated for updated packages:
https://github.com/deskflow/deskflow/blob/80488033f75868573c0b2023920a30d931ad57fa/.github/workflows/ci.yml#L340

**Comment by nbolton** [comment]
Ah, I need to create `DEPLOY_TOKEN` secret. 

@sithlord48 I think we should instead only deploy to winget for stable release instead of continuous. What do you think?

**Comment by nicola02nb** [comment]
@nbolton I used a `DEPLOY_TOKEN` that was already present into CI, maybe you should create a different one(it's the first time I do something like this... I'm kind a newebie)

**Comment by nicola02nb** [comment]
@nbolton Anyway I think you should wait for first approval to the package into [winget-pkgs](https://github.com/microsoft/winget-pkgs) repo (https://github.com/microsoft/winget-pkgs/pull/185685) then you can start using this CI.

It should take less than 24h if everithing goes well.

**Comment by nbolton** [comment]
> @nbolton I used a `DEPLOY_TOKEN` that was already present into CI, maybe you should create a different one(it's the first time I do something like this... I'm kind a newebie)

I misspoke, I was meant to say `WINGET_DEPLOY_TOKEN`. It'd make sense to make a new token for this.

I'll make some suggestions.

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -327,3 +327,20 @@ jobs:
...
+  winget:
+    needs: release
+    name: Publish winget package
+    runs-on: windows-latest
+    steps:
+      - name: Submit package to Windows Package Manager Community Repository
+        run: |
+
+          $packageId = "Deskflow.Deskflow(64-bit)"
+          $gitToken = "${{ secrets.DEPLOY_TOKEN }}"
```
Comment (nbolton) [inline .github/workflows/ci.yml:340]:
```suggestion
          $gitToken = "${{ secrets.WINGET_DEPLOY_TOKEN }}"
```

**Comment by nicola02nb** [comment]
> > @nbolton I used a `DEPLOY_TOKEN` that was already present into CI, maybe you should create a different one(it's the first time I do something like this... I'm kind a newebie)
> 
> I misspoke, I was meant to say `WINGET_DEPLOY_TOKEN`. It'd make sense to make a new token for this.
> 
> I'll make some suggestions.
> 
> I approved the run to confirm that it would be skipped (which it was, since release is only run on master and tags, not PRs).

Ok, cool.

If you need anything just let me know

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -327,3 +327,20 @@ jobs:
           prerelease: false
           files: |
             package-*/*
+            
+  winget:
+    needs: release
+    name: Publish winget package
+    runs-on: windows-latest
```
Comment (nbolton) [inline .github/workflows/ci.yml:334]:
```suggestion
  winget-publish:
    needs: release
    if: contains(github.ref, '/tags/v')
    runs-on: windows-latest
```

For consistency with other jobs, I removed `name` (not that it's wrong or anything, it's because I like to be consistent with the other jobs... one of my idiosyncracies).

I also added an `if` so it doesn't deploy the continuous builds which are highly unstable and break often.

I think we should use a unique `WINGET_DEPLOY_TOKEN` because then if the token is leaked (not that I don't trust Microsoft or anything), we know how it happened.

**Comment by nbolton** [comment]
> I think we should use a unique WINGET_DEPLOY_TOKEN because then if the token is leaked (not that I don't trust Microsoft or anything), we know how it happened.

Doing a bit of digging to find out what permissions this needs, and found the docs:
- [`winget-create/doc/token.md`](https://github.com/microsoft/winget-create/blob/main/doc/token.md)
- [GitHub Personal Access Token (classic) Permissions
](https://github.com/microsoft/winget-create/blob/main/README.md#github-personal-access-token-classic-permissions)

Saving here for visibility and posterity :

> The token command of the [Winget-Create](https://github.com/microsoft/winget-create/blob/main/README.md) tool is designed to manage cached GitHub personal access tokens used by the tool for interacting with the [Windows Package Manager repo](https://docs.microsoft.com/windows/package-manager/) automatically. To use the token command, you can specify whether you want to store a new [GitHub token](https://docs.github.com/en/github/authenticating-to-github/creating-a-personal-access-token) or clear any existing cached tokens. If you choose not to provide a token when storing, the tool will initiate an OAuth flow and prompt for your GitHub login credentials.

> GitHub personal access token used for direct submission to the Windows Package Manager repo. If no token is provided, tool will prompt for GitHub login credentials.

> When [creating your own GitHub Personal Access Token (PAT)](https://docs.github.com/en/github/authenticating-to-github/keeping-your-account-and-data-secure/creating-a-personal-access-token) to be used with WingetCreate, make sure the following permissions are selected.
> Select the public_repo scope to allow access to public repositories: `public_repo scope`
> (Optional) Select the delete_repo scope permission if you want WingetCreate to automatically delete the forked repo that it created if the PR submission fails.

image

File: .github/workflows/ci.yml (hunk 1 above)
Comment (nbolton) [reply .github/workflows/ci.yml:340]: I have created this token, which grants public repo access so Microsoft can download the package from us.

**Review by nbolton — CHANGES_REQUESTED** [review]

File: .github/workflows/ci.yml (hunk 2 above)
Comment (nicola02nb) [reply .github/workflows/ci.yml:334]: Yeah, it's only for security reson, I agree

Comment (sithlord48) [reply .github/workflows/ci.yml:334]: This also needs to be guarded so it only happens on windows 
Something like 
`if: ((runner.os == 'Windows') && (contains(github.ref, 'tags/v')))`

Comment (nbolton) [reply .github/workflows/ci.yml:334]: This conversation was resolved but no change was made. Reopening.

File: .github/workflows/ci.yml (hunk 1 above)
Comment (nbolton) [reply .github/workflows/ci.yml:340]: This conversation was resolved but no change was made. Reopening.

File: .github/workflows/ci.yml (hunk 2 above)
Comment (nicola02nb) [reply .github/workflows/ci.yml]: Sorry, forgot to commit

**Comment by nbolton** [comment]
:bulb: We make this into a composite action so that we can test it with a `workflow_dispatch` before landing this PR into master... otherwise it's a a blind faith land with some trial and error (an approach which I am against).

**Comment by nicola02nb** [comment]
> > @nicola02nb ~Ah, I need to create `DEPLOY_TOKEN` secret.~
> > @sithlord48 I think we should instead only deploy to winget for stable release instead of continuous. What do you think?
> 
> Does winget have two release streams ? if not then yeah we should only have the releases here

As far as I know it has only one release stream, but you can create two manifests with differents `PackageIdentifier`, with different download link, if I'm not wrong....

**Comment by marlop352** [comment]
> @nicola02nb ~Ah, I need to create `DEPLOY_TOKEN` secret.~
> 
> @sithlord48 I think we should instead only deploy to winget for stable release instead of continuous. What do you think?

on a related note, can the continuous build be marked as "Pre-release" here on GitHub since it's not considered stable?

**Comment by nicola02nb** [comment]
> > @nicola02nb ~Ah, I need to create `DEPLOY_TOKEN` secret.~
> > @sithlord48 I think we should instead only deploy to winget for stable release instead of continuous. What do you think?
> 
> on a related note, can the continuous build be marked as "Pre-release" here on GitHub since it's not considered stable?



> 💡 We make this into a composite action so that we can test it with a `workflow_dispatch` before landing this PR into master... otherwise we'll have to do a blind faith land with some trial and error (an approach which I am against).

Ok, it's defenetly  better of doing a bunch of random commits and hoping it works....

Should I do it, or I'll let you the job?

**Comment by sithlord48** [comment]
> > @nicola02nb ~Ah, I need to create `DEPLOY_TOKEN` secret.~
> > @sithlord48 I think we should instead only deploy to winget for stable release instead of continuous. What do you think?
> 
> on a related note, can the continuous build be marked as "Pre-release" here on GitHub since it's not considered stable?

its not considered not stable and we want ppl not using system packages to test this for us

**Comment by sithlord48** [comment]
go for the composite action

**Comment by marlop352** [comment]
> > > @nicola02nb ~Ah, I need to create `DEPLOY_TOKEN` secret.~
> > > @sithlord48 I think we should instead only deploy to winget for stable release instead of continuous. What do you think?
> > 
> > 
> > on a related note, can the continuous build be marked as "Pre-release" here on GitHub since it's not considered stable?
> 
> its not considered not stable and we want ppl not using system packages to test this for us

well, that is fairly contradictory, Mr Bolton said: "I think we should instead only deploy to winget for stable release instead of continuous." so continuous is *not* the stable release, also you just said "we want ppl not using system packages to test this for us" another clear indicator that continuous is not to be considered a stable release.

If you won't mark it as pre-release as it should for the sake of "we want more people to test it" then it better be indicated on the release text that continuous is not the project's stable release.

**Comment by sithlord48** [comment]
> > > > @nicola02nb ~Ah, I need to create `DEPLOY_TOKEN` secret.~
> > > > @sithlord48 I think we should instead only deploy to winget for stable release instead of continuous. What do you think?
> > > 
> > > 
> > > on a related note, can the continuous build be marked as "Pre-release" here on GitHub since it's not considered stable?
> > 
> > 
> > its not considered not stable and we want ppl not using system packages to test this for us
> 
> well, that is fairly contradictory, Mr Bolton said: "I think we should instead only deploy to winget for stable release instead of continuous." so continuous is _not_ the stable release, also you just said "we want ppl not using system packages to test this for us" another clear indicator that continuous is not to be considered a stable release.
> 
> If you won't mark it as pre-release as it should for the sake of "we want more people to test it" then it better be indicated on the release text that continuous is not the project's stable release.

I would not expect any one but us to keep up w/ the continuous release we could have several a day and while it is not inherently unstable winget or any other package provider is not going to want us updating our package that may times  a day .

**Comment by nbolton** [comment]
>> another clear indicator that continuous is not to be considered a stable release.
> it is not inherently unstable

Perhaps of interest to someone out there: I believe that the observation made by @marlop352 is called a _false dichotomy_, which @sithlord48 correctly addressed. The false dichotomy here occurs when someone assumes that naming one object "stable" means the other must be "unstable," creating a binary choice between the two. This limits the possibilities to only two—stable vs. unstable—even if other options are available. In reality, the second object could be stable, unstable, neither, or have a different characteristic entirely.

However, I can see how many people would apply this bias subconsciously :slightly_smiling_face:

**Comment by nbolton** [comment]
> go for the composite action

@nicola02nb Want me to handle this or do you want to take a stab at it? :)

**Comment by nbolton** [comment]
> > > go for the composite action
> > 
> > 
> > @nicola02nb Want me to handle this or do you want to take a stab at it? :)
> 
> I have already done something in last commit, but I fell like you should do it, you're probably more expert than me 😅

Oh look at that, so you did. All you need to do is call it. I'll create a `workflow_dispatch` in `master` so we can test it before landing the PR.

**Comment by sithlord48** [comment]
Before this can land we need to tidy up this commit history.

File: .github/actions/winget-publish/action.yaml [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+  installer-url-template:
+    description: "Template for installer URL. Use {version} as placeholder"
+    required: true
+    default: "https://github.com/deskflow/deskflow/releases/download/v{version}/deskflow_{version}_windows_x64.msi"
+
+runs:
+  using: "composite"
+  steps:
+    - name: Submit package to Windows Package Manager Community Repository
+      if: runner.os == 'Windows'
```
Comment (sithlord48) [inline .github/actions/winget-publish/action.yaml:24]: pretty sure this needs to be 
`${{ runner.os == 'Windows' }}`

**Review by marlop352 — CHANGES_REQUESTED** [review]

File: .github/workflows/ci.yml [hunk 4]
```diff
@@ -327,3 +327,20 @@ jobs:
...
+    steps:
+      - name: Submit package to Windows Package Manager Community Repository
+        run: |
+
+          $packageId = "Deskflow.Deskflow(64-bit)"
+          $gitToken = "${{ secrets.WINGET_DEPLOY_TOKEN }}"
+          
+          $packageVersion = "${{env.DESKFLOW_VERSION}}"
+          $installerUrl = "https://github.com/deskflow/deskflow/releases/download/continuous/deskflow_${packageVersion}_windows_x64.msi"
```
[the remark this answers is not in the document]
Comment (sithlord48) [reply .github/workflows/ci.yml:343]: Correct the link should be 

`          $installerUrl = "https://github.com/deskflow/deskflow/releases/download/v${packageVersion}/deskflow_${packageVersion}_windows_x64.msi"`

It will not work for 1.17.0 as it has files with a +r1 added to the tag for the version .. it will work with 1.18.0+ 
unless @nbolton  we rename the dl files to match the tag

[the remark this answers is not in the document]
Comment (sithlord48) [reply .github/workflows/ci.yml:343]: Please note the "version number" for a tag of V 1.18.0 will be `1.18.0.0 for the file name i've added the .0 the the link above as we only plan to push tagged versions

File: .github/workflows/ci.yml [hunk 5]
```diff
@@ -327,3 +327,16 @@ jobs:
           prerelease: false
           files: |
             package-*/*
+            
+  winget-publish:
+    needs: release
+    if: (runner.os == 'Windows') && contains(github.ref, 'tags/v')
+    runs-on: windows-latest
+    steps:
+      - name: Submit package to Windows Package Manager Community Repository
```
Comment (nbolton) [inline .github/workflows/ci.yml:336]:
```suggestion
      - name: Submit
```

I know this seems picky, but long names can be a bit of a distraction in the workflow summary.

File: .github/workflows/ci.yml [hunk 6]
```diff
@@ -327,3 +327,16 @@ jobs:
...
+        uses: ./.github/actions/winget-publish
+        with:
+          release-version: ${{env.DESKFLOW_VERSION}}
+          package-id: 'Deskflow.Deskflow(64-bit)'
```
Comment (nbolton) [inline .github/workflows/ci.yml:340]:
```suggestion
          package-id: 'Deskflow.Deskflow'
```

I wonder if 64-bit is redundant, since we don't offer a 32-bit MSI.

File: .github/workflows/ci.yml [hunk 7]
```diff
@@ -327,3 +327,16 @@ jobs:
           prerelease: false
           files: |
             package-*/*
+            
+  winget-publish:
+    needs: release
+    if: (runner.os == 'Windows') && contains(github.ref, 'tags/v')
```
Comment (nbolton) [inline .github/workflows/ci.yml:333]:
```suggestion
    if: contains(github.ref, 'tags/v')
```

I realised after thinking about this that `(runner.os == 'Windows')` is redundant because of `runs-on: windows-latest`

File: .github/workflows/ci.yml [hunk 8]
```diff
@@ -327,3 +327,16 @@ jobs:
...
+    runs-on: windows-latest
+    steps:
+      - name: Submit package to Windows Package Manager Community Repository
+        uses: ./.github/actions/winget-publish
```
Comment (nbolton) [inline .github/workflows/ci.yml:337]: Now that we're using a composite, we need to do checkout:
```
      # Fancy checkout gets all the tags
      # it also makes sure we can use git --describe correctly
      - name: Fancy Checkout
        uses: sithlord48/fancy-checkout@v1.0.0
```

File: .github/workflows/ci.yml [hunk 9]
```diff
@@ -327,3 +327,16 @@ jobs:
...
+        with:
+          release-version: ${{env.DESKFLOW_VERSION}}
```
Comment (nbolton) [inline .github/workflows/ci.yml:339]: This env var depends on another action, which must go after checkout:
```
      - name: Get version
        uses: ./.github/actions/get-version
```

File: .github/actions/winget-publish/action.yaml [hunk 10]
```diff
@@ -0,0 +1,40 @@
...
+inputs:
+  release-version:
+    description: "Version to publish to Winget package manager (without 'v' prefix)"
+    required: true
+  package-id:
+    description: "Package identifier in Winget repository"
+    required: true
+    default: "Deskflow.Deskflow(64-bit)"
+  token:
+    description: "GitHub token with permissions to submit to Winget repository"
```
Comment (nbolton) [inline .github/actions/winget-publish/action.yaml:13]:
```suggestion
    description: "GitHub token with public read permissions on the source repo"
```

AFAIK, the token only needs permission to read from the Deskflow repo.

File: .github/actions/winget-publish/action.yaml [hunk 11]
```diff
@@ -0,0 +1,40 @@
...
+description: A composite action to publish packages to the Windows Package Manager (Winget) repository
+
+inputs:
+  release-version:
+    description: "Version to publish to Winget package manager (without 'v' prefix)"
+    required: true
+  package-id:
+    description: "Package identifier in Winget repository"
+    required: true
+    default: "Deskflow.Deskflow(64-bit)"
```
Comment (nbolton) [inline .github/actions/winget-publish/action.yaml:11]:
```suggestion
    default: "Deskflow.Deskflow"
```

I wonder if 64-bit is redundant, since we don't offer a 32-bit MSI.

File: .github/workflows/ci.yml (hunk 9 above)
Comment (nbolton) [reply .github/workflows/ci.yml:339]: Actually, I think that this could go inside the composite action. Actions can call other actions no problem. No need to pass it as a parameter.

File: .github/workflows/ci.yml [hunk 12]
```diff
@@ -327,3 +327,16 @@ jobs:
...
+          package-id: 'Deskflow.Deskflow(64-bit)'
+          token: ${{ secrets.WINGET_DEPLOY_TOKEN }}
+          installer-url-template: 'https://github.com/deskflow/deskflow/releases/download/v{version}/deskflow_{version}.0_windows_x64.msi'
```
Comment (nbolton) [inline .github/workflows/ci.yml:342]: I think perhaps this could be embedded in the composite action. No need to have this as a parameter.

File: .github/actions/winget-publish/action.yaml [hunk 13]
```diff
@@ -0,0 +1,40 @@
...
+    - name: Submit package to Windows Package Manager Community Repository
+      if: ${{ runner.os == 'Windows' }}
+      env:
+        GITHUB_TOKEN: ${{ inputs.token }}
+      run: |
+        # Download latest wingetcreate
+        Invoke-WebRequest https://aka.ms/wingetcreate/latest -OutFile wingetcreate.exe
+        
+        # Replace {version} in template with actual version
+        $installerUrl = "${{ inputs.installer-url-template }}".Replace("{version}", "${{ inputs.release-version }}")
```
Comment (nbolton) [inline .github/actions/winget-publish/action.yaml:32]: As mentioned in my other comment, let's not have this as a param. Embedding it in the PowerShell script should be fine, that way you don't need the `Replace` logic.

**Review by nbolton — CHANGES_REQUESTED** [review]
A lot of comments, might be a bit hard to follow my reasoning. I wonder if we should discuss on Matrix? https://matrix.to/#/#deskflow-dev:matrix.org

File: .github/workflows/ci.yml (hunk 6 above)
Comment (nbolton) [inline .github/workflows/ci.yml:340]:
```suggestion
```

I'm not sure this needs to be a param.

File: .github/workflows/ci.yml (hunk 12 above)
Comment (nbolton) [inline .github/workflows/ci.yml:342]:
```suggestion
```
I think perhaps this could be embedded in the composite action. No need to have this as a parameter.

**Comment by sithlord48** [comment]
The commit count just keeps growing.. 
Please fix the commits instead of adding new ones. 
LMK if your unable to and ill do if for you .

**Comment by marlop352** [comment]
> > > another clear indicator that continuous is not to be considered a stable release.
> > 
> > 
> > it is not inherently unstable
> 
> Perhaps of interest to someone out there: I believe that the observation made by @marlop352 is called a _false dichotomy_, which @sithlord48 correctly addressed. The false dichotomy here occurs when someone assumes that naming one object "stable" means the other must be "unstable," creating a binary choice between the two. This limits the possibilities to only two—stable vs. unstable—even if other options are available. In reality, the second object could be stable, unstable, neither, or have a different characteristic entirely.
> 
> However, I can see how many people would apply this bias subconsciously, and even I made this same misinterpretation by raising PR #7813 (which we decided not to land).

Well @nbolton , you said:
>I also added an if so it doesn't deploy the continuous builds which are highly unstable and break often.

while reviewing this very PR. I had dropped the subject before I saw that comment but you need to decide if continuous is stable or not and stick to whatever narrative you choose. (I know the change I requested is not happening because the team doesn't want it to be done, this is just to point out that you are being **very** contradictory.)

I'm gonna solve my particular problem automating installs and stuff on windows by just using the winget package whenever it's ready as it won't have continuous builds on it. As for Linux I'm just gonna hope the maintainers are sane and point them here if they think shipping continuous is a good idea.

(No need to reply to this unless by a miracle the team changed it's mind on the matter)

**Comment by marlop352** [comment]
@sithlord48 I get that you all want this to be a single commit, but I suggest to only squash and force-push it after the code is considered ready.

**Comment by nbolton** [comment]
> Well @nbolton , you said:
> > I also added an if so it doesn't deploy the continuous builds which are highly unstable and break often.

Ah, my mistake! I meant to say:

"I also added an if so it doesn't deploy the continuous builds which [_can be_] highly unstable and [_may_] break often."

Sorry about the confusion :slightly_smiling_face:

**Review by sithlord48 — APPROVED** [review]
LGTM

**Comment by nbolton** [comment]
> LGTM

Will test before landing on #7822: `winget-test`

**Review by nbolton — COMMENTED** [review]
I'm seeing an error when testing in PR #7822, but I guess that's because it was already submitted?
```
Telemetry Settings
------------------
The Windows Package Manager Manifest Creator collects usage data in order to improve your experience.
The data is anonymous and collected only by Microsoft.
By default, telemetry is enabled but can be disabled by running `wingetcreate settings` and editing your settings file.
Retrieving latest manifest for Deskflow.Deskflow
Downloading and parsing: https://github.com/deskflow/deskflow/releases/download/v1.17.0/deskflow-1.17.0.0_win64.msi...
Could not populate manifest metadata through GitHub's API.
Generating a preview of your manifests...
Version manifest preview:
# Created using wingetcreate 1.6.5.0
# yaml-language-server: $schema=https://aka.ms/winget-manifest.version.1.6.0.schema.json
PackageIdentifier: Deskflow.Deskflow
PackageVersion: 1.17.0
DefaultLocale: en-US
ManifestType: version
ManifestVersion: 1.6.0
Installer manifest preview:
# Created using wingetcreate 1.6.5.0
# yaml-language-server: $schema=https://aka.ms/winget-manifest.installer.1.6.0.schema.json
PackageIdentifier: Deskflow.Deskflow
PackageVersion: 1.17.0
InstallerLocale: en-US
InstallerType: wix
ProductCode: '{0D79F5CB-9C83-4BCD-9500-19E369BE6303}'
Installers:
- Architecture: x64
  InstallerUrl: https://github.com/deskflow/deskflow/releases/download/v1.17.0/deskflow-1.17.0.0_win64.msi
  InstallerSha[25](https://github.com/deskflow/deskflow/actions/runs/11592983993/job/32276259251#step:3:27)6: F70A21F8E4F415B47AD34AA30BDDC8D929AD42DBAEE071CBED40D577BD1E0F49
ManifestType: installer
ManifestVersion: 1.6.0
Default locale manifest preview:
# Created using wingetcreate 1.6.5.0
# yaml-language-server: $schema=https://aka.ms/winget-manifest.defaultLocale.1.6.0.schema.json
PackageIdentifier: Deskflow.Deskflow
PackageVersion: 1.17.0
PackageLocale: en-US
Publisher: Deskflow
PublisherUrl: https://github.com/deskflow
PublisherSupportUrl: https://github.com/deskflow/deskflow/issues
PackageName: Deskflow
PackageUrl: https://github.com/deskflow/deskflow
License: GPL-2.0
ShortDescription: Deskflow lets you share one mouse and keyboard between multiple computers on Windows, macOS and Linux. It's like a software KVM (but without video).
Tags:
- keyboard
- keyboard-emulation
- mouse
- mouse-emulation
- network
Documentations:
- DocumentLabel: Wiki
  DocumentUrl: https://github.com/deskflow/deskflow/wiki
ManifestType: defaultLocale
ManifestVersion: 1.6.0
Manifest saved to D:\a\deskflow\deskflow\manifests\d\Deskflow\Deskflow\1.17.0
Manifest validation succeeded: True
Submitting a manifest without any updated changes is not allowed. 
Please ensure that the updated manifest you are submitting is different from the existing package.
Error: Process completed with exit code 1.
```

**Comment by marlop352** [comment]
> Perhaps a red herring, as I guess that's because it was already submitted:

It was submitted and approved here: https://github.com/microsoft/winget-pkgs/pull/186364

That error should be expected since there is no new release yet.

I never used wingetcreate, is the line `Could not populate manifest metadata through GitHub's API.` from the test-action logs normal?

**Comment by nicola02nb** [comment]
Just tried from cli on my pc and it gave sht same error, so the version you are sending should be newer...

**Comment by nbolton** [comment]
> It was submitted and approved here: [microsoft/winget-pkgs#186364](https://github.com/microsoft/winget-pkgs/pull/186364)

Ah, so it looks like they pulled it from the URL in the manifest. We can't update because there is no change to the manifest I believe. Since they're pulling it... I wonder why we would need to push.

**Comment by marlop352** [comment]
> > It was submitted and approved here: [microsoft/winget-pkgs#186364](https://github.com/microsoft/winget-pkgs/pull/186364)
> 
> Ah, so it looks like they pulled it from the URL in the manifest. We can't update because there is no change to the manifest I believe. Since they're pulling it... I wonder why we would need to push.

Well, you would need a manual PR every time for them to pull, the PR to add was done by [nicola02nb](https://github.com/nicola02nb) who also started this PR.

**Comment by nbolton** [comment]
So will the [URL in the manifest](https://github.com/nicola02nb/winget-pkgs/blob/bd7404d71d9248eb707fbcddc3f6c799afb92618/manifests/d/Deskflow/Deskflow/1.17.0/Deskflow.Deskflow.installer.yaml#L11) ever be updated or is that just a bootstrapping thing and will always point to the first MSI that was submitted? I've asked the question to @stephengillie: https://github.com/microsoft/winget-pkgs/pull/186364#issuecomment-2447109729

**Review by nbolton — APPROVED** [review]
@sithlord48 Approving. I was unable to test, since Microsoft pulled the MSI from us as part of adopting the first manifest. I guess we'll only know if it works when we eventually tag v1.18. Hopefully it works, otherwise it's going to cause some confusion when we tag.

**Review by marlop352 — APPROVED** [review]

**Comment by marlop352** [comment]
> So will the [URL in the manifest](https://github.com/nicola02nb/winget-pkgs/blob/bd7404d71d9248eb707fbcddc3f6c799afb92618/manifests/d/Deskflow/Deskflow/1.17.0/Deskflow.Deskflow.installer.yaml#L11) ever be updated or is that just a bootstrapping thing and will always point to the first MSI that was submitted? I've asked the question to @stephengillie: [microsoft/winget-pkgs#186364 (comment)](https://github.com/microsoft/winget-pkgs/pull/186364#issuecomment-2447109729)

[nicola02nb](https://github.com/nicola02nb) volunteered to make the PR updating the manifest: https://github.com/microsoft/winget-pkgs/pull/186364#discussion_r1822617794

**Comment by marlop352** [comment]
@nbolton on a related note, it would be good if the naming of the release artifacts was more consistent
image
some have "-" and some have "_".

Also the format that is currently being used for the windows msi file name is different from the one on this PR but +/- equal to the one already on the manifest that was merged in the [microsoft/winget-pkgs](https://www.github.com/microsoft/winget-pkgs) repo: https://github.com/microsoft/winget-pkgs/blob/master/manifests/d/Deskflow/Deskflow/1.17.0/Deskflow.Deskflow.installer.yaml

**Comment by sithlord48** [comment]
Yes you can tell what is made by cpack and what is not the - ones are done by cpack the _ ones are yet to be moved into cpack

**Comment by marlop352** [comment]
> Yes you can tell what is made by cpack and what is not the - ones are done by cpack the _ ones are yet to be moved into cpack

So if I understood correctly all of them are being moved to cpack and that will be finished before the next tagged release?

**Comment by marlop352** [comment]
In this case I would leave the current manifest that is already at the winget repo alone and keep as artifact in the 1.17.0 release only the msi that follows the "old" naming syntax.

**Comment by sithlord48** [comment]
> > Yes you can tell what is made by cpack and what is not the - ones are done by cpack the _ ones are yet to be moved into cpack
> 
> So if I understood correctly all of them are being moved to cpack and that will be finished before the next tagged release?

Yes that is the plan , we will retroactively rename the 1.17 files and at some point the continuous will align and eventually it will become 1.18.0. our links should be consistent by then. there is a pr to fix the  manifest on the winget side to use the new scheme .

**Comment by marlop352** [comment]
> > > Yes you can tell what is made by cpack and what is not the - ones are done by cpack the _ ones are yet to be moved into cpack
> > 
> > 
> > So if I understood correctly all of them are being moved to cpack and that will be finished before the next tagged release?
> 
> Yes that is the plan , we will retroactively rename the 1.17 files and at some point the continuous will align and eventually it will become 1.18.0. our links should be consistent by then. there is a pr to fix the manifest on the winget side to use the new scheme .

I would recommend against renaming the 1.17 release as that would break a fair amount of automations, including the AUR packages.

**Comment by sithlord48** [comment]
> > > > Yes you can tell what is made by cpack and what is not the - ones are done by cpack the _ ones are yet to be moved into cpack
> > > 
> > > 
> > > So if I understood correctly all of them are being moved to cpack and that will be finished before the next tagged release?
> > 
> > 
> > Yes that is the plan , we will retroactively rename the 1.17 files and at some point the continuous will align and eventually it will become 1.18.0. our links should be consistent by then. there is a pr to fix the manifest on the winget side to use the new scheme .
> 
> I would recommend against renaming the 1.17 release as that would break a fair amount of automations, including the AUR packages.

This is a necessary evil and should be the last time we need to worry about that . I will go make a comment on the -bin package on the aur.

**Comment by marlop352** [comment]
> > I would recommend against renaming the 1.17 release as that would break a fair amount of automations, including the AUR packages.
> 
> This is a necessary evil and should be the last time we need to worry about that . I will go make a comment on the -bin package on the aur.

I see no reason for that to be necessary other than that you want to do so.
Why would you need to rename an already released artifact? (other than the name being incorrect, but that is not the case here)

**Comment by SelfRef** [comment]
I agree, there's no really a reason to rename existing binaries versions to new schema/link after they were released. The new naming schema can be always applied while creating new releases, it can be even a patch one like 1.17.1 just for the sake of creating a new version.

I use semi-automated system to notify me about updates, but it doesn't take into account that the existing release may change, especially 3 weeks later. Not a big deal really, just unnecessary.

**Comment by nbolton** [comment]
Perhaps it might be useful to contribute about version numbers on discussion: #7843
